crates/bevy_ecs/src/term_query/builder.rs

```diff
@@ -0,0 +1,84 @@
+use std::marker::PhantomData;
+
+use crate::{
+    component::ComponentId,
+    prelude::{Component, With, Without, World},
+};
+
+use super::{ComponentTerm, QueryTermGroup, Term, TermQueryState, TermVec};
+
+pub struct QueryBuilder<'w, Q: QueryTermGroup = ()> {
+    terms: TermVec<Term>,
+    current_term: usize,
+    world: &'w mut World,
+    _marker: PhantomData<Q>,
+}
+
+impl<'w, Q: QueryTermGroup> QueryBuilder<'w, Q> {
+    pub fn new(world: &'w mut World) -> Self {
+        let mut terms = TermVec::new();
+        Q::init_terms(world, &mut terms);
+        Self {
+            current_term: terms.len(),
+            terms,
+            world,
+            _marker: PhantomData::default(),
+        }
+    }
+
+    pub fn term<T: QueryTermGroup>(&mut self) -> &mut Self {
+        T::init_terms(self.world, &mut self.terms);
+        self
+    }
+
+    pub unsafe fn term_at(&mut self, index: usize) -> &mut Self {
```

Comment:
We should think carefully about whether this method should be `pub`. It seems *very* easy to screw up, and it's not clear how an external library would make use of it.

Reply:
Something like this method is needed in order to allow users to set the target component for a `Ptr` term. Although since that's its only safe use case maybe we just add something more specialized. 

Debateably we don't even need a Ptr term since you can always iterate the raw fetched terms if you want to anyway but it is a lot more ergonomic for some rare use cases and the abstraction is very thin.